seq: use augmenting addition to update value

jfinkels opened the pull request:
This pull request fixes a floating-point precision issue with the update in the main loop of `seq`. Before this commit, the behavior was

    $ seq 0.1 -0.1 -0.2
    0.1
    0.0
    -0.1

which was incorrect: -0.2 should be printed as well. This was happening
because the update in the main loop was essentially

    value = start + (i * increment);

which caused `value` to become `-0.20000000000000004`. That exceeds the lower bound of -0.2, so that number was not printed. By changing the update to

    value += increment;

we eliminate this issue.

tertsdiepraam commented:
Very interesting! This does indeed mitigate the issue, but it's not quite solved yet. I just tried:
```
seq 0 0.1 10000.1
```
and GNU includes `10000.1` while we (including this PR) don't.

I checked out the FreeBSD implementation and they use the multiplication method with an additional check for the last value:
```C
/*
* Did we miss the last value of the range in the loop above?
*
* We might have, so check if the printable version of the last
* computed value ('cur') and desired 'last' value are equal.  If they
* are equal after formatting truncation, but 'cur' and
* 'last_shown_value' are not equal, it means the exit condition of the
* loop held true due to a rounding error and we still need to print
* 'last'.
*/
asprintf(&cur_print, fmt, cur);
asprintf(&last_print, fmt, last);
if (strcmp(cur_print, last_print) == 0 && cur != last_shown_value) {
    fputs(last_print, stdout);
    fputs(sep, stdout);
}
```

miDeb commented:
AFAIK gnu uses `long double`s (80 bit floats on x86), which are not available in rust.

jfinkels commented:
So maybe the FreeBSD approach can offer a bit more protection, but we may not be completely safe against this type of incompatibility.

blyxxyz commented:
Arbitrary-precision decimal arithmetic might be ideal for correctness, because all the input is decimal to begin with. Maybe with [this crate](https://lib.rs/crates/bigdecimal).

That would make it trickier to support the `--format` option for printf-style float formatting, but numbers could be converted to floats in that case.

jfinkels commented on src/uu/seq/src/seq.rs:
```diff
@@ -102,24 +130,28 @@ impl FromStr for Number {
                 }
             }
-            Err(_) => match s.parse::<f64>() {
-                Ok(value) if value.is_nan() => Err(format!(
-                    "invalid 'not-a-number' argument: '{}'\nTry '{} --help' for more information.",
-                    s,
-                    uucore::execution_phrase(),
-                )),
```
Removing this arm of the match statement has caused a regression in the test `test_seq::test_rejects_nan`. We need to check here if the parsed value is `NaN`, but the `bigdecimal::BigDecimal` struct doesn't support NaN it seems: https://github.com/akubera/bigdecimal-rs/issues/67 . So I could add an additional `parse::<f64>()` just to check for NaN? Suggestions welcome.

blyxxyz replied:
Parsing a `f64` sounds like a great solution, it seems to catch all the formats GNU seq/`strtod` catches.

You could then watch out for inf/-inf too. `seq inf` counts up endlessly, which might be useful enough to support. `seq -- -inf 0` just outputs `-inf` endlessly, that's less useful. (And `seq -- inf -1 0` outputs `inf` endlessly.)

jfinkels replied:
I'll create a new issue to represent the feature request of supporting `inf` and `-inf` arguments.

jfinkels replied:
Oh, I see what you're saying. You're saying that the `uutils` version of `seq` on the `master` branch already supports `seq inf` and `seq -- -inf 0`, and this pull request as it is today will cause those invocations to no longer work. There are no test cases for `inf` and `-inf` arguments, so I guess I need to add them now to prevent further regressions. Okay, I'll do that too.

blyxxyz replied:
GNU `seq` also supports `inf`, but you're right that it would be a local regression, I didn't even think of that.

jfinkels commented:
I have updated this pull request to use the `bigdecimal` crate. However, I've introduced a regression in the test for NaN arguments. Suggestions welcome, see my line comment.

jfinkels commented:
I just updated this pull request with support for the NaN error message, but @blyxxyz pointed out that there are some further use cases (namely, `inf` and `-inf` arguments) that will see a regression with the code as it is today. I'll add more test cases to cover those situations and I'll ensure that the changes I propose don't reduce the functionality of `seq`.

jfinkels commented:
Pull request #2647 adds two missing unit test cases that witness our support of `seq 0 inf` and `seq -inf 0`. I'll include those changes in this branch in order to ensure that this pull request does not cause a regression.

jfinkels commented:
Hmm... I'm not sure how to do this without having a `BigDecimal` implementation that also supports positive and negative infinity.

blyxxyz commented on src/uu/seq/src/seq.rs:
```diff
@@ -268,38 +200,64 @@ fn done_printing<T: Num + PartialOrd>(next: &T, increment: &T, last: &T) -> bool
     }
-    if (first >= last && increment < 0f64) || (first <= last && increment > 0f64) {
-        print!("{}", terminator);
+    if (first >= last && increment < BigDecimal::zero())
+        || (first <= last && increment > BigDecimal::zero())
+    {
```
Would this work?
```suggestion
    if !is_first_iteration {
```

jfinkels replied:
I don't think so. It looks like this block of code is supposed to be executed only after the very last item is printed, so I guess it should really be `if (value >= last && ...) || (value <= last && ...)`. However, I don't know why this is even an option; there is no `-t` option in the GNU coreutils version of `seq`: https://www.gnu.org/software/coreutils/manual/html_node/seq-invocation .

I will create a new pull request where I suggest we drop the functionality altogether.

jfinkels replied:
https://github.com/uutils/coreutils/pull/2677

blyxxyz replied:
It runs after the very last item is printed, but only if there was an item at all. There's a similar case in `print_seq_integers`.

`if !is_first_iteration` works because if there weren't any items `is_first_iteration` will still be true. (And it's outside the loop, so it'll only run at the end.)

I'm not 100% sure about removing it, and I don't think "it's not in GNU" is always a sufficient reason, but I guess I don't think I'd use it myself.

blyxxyz commented on src/uu/seq/src/seq.rs:
```diff
@@ -268,38 +200,64 @@ fn done_printing<T: Num + PartialOrd>(next: &T, increment: &T, last: &T) -> bool
-        if pad && before_dec < padding {
-            for _ in 0..(padding - before_dec) {
-                print!("0");
+        if pad && before_dec < width {
+            for _ in 0..(width - before_dec) {
+                write!(stdout, "0")?;
             }
         }
-        print!("{}", istr);
-        i += 1;
-        value = first + i as f64 * increment;
+        write!(stdout, "{}", istr)?;
```
This seems to do the trick:
```suggestion
        write!(
            stdout,
            "{value:>0width$.precision$}",
            value = value,
            width = if pad { width + 1 + largest_dec } else { 0 },
            precision = largest_dec
        )?;
```

jfinkels replied:
I'll put your suggested change in a separate pull request.

jfinkels replied:
https://github.com/uutils/coreutils/pull/2676

blyxxyz commented on src/uu/seq/src/seq.rs:
```diff
@@ -211,16 +121,38 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
                 false,
             )
         }
+        (Number::MinusZeroFloat, last, increment) => print_seq(
```
This should also cover `MinusZeroInt`

jfinkels replied:
Yes, I'll add a test case for that as well. Thanks.

blyxxyz reviewed (changes requested):
I think changing `RangeBigDecimal` into `(BigDecimal, BigDecimal, Option<BigDecimal>)` could work. With something like this logic:
- If the start or increment is `inf` or `-inf`, abort. GNU seq supports it but it's not really useful for anything.
- If the end is `inf` and the increment is positive or the end is `-inf` and the increment is negative set the end to `None` and just loop endlessly.
- If the end is `-inf` and the increment is negative or the end is `inf` and the increment is positive, stop immediately without any output.

jfinkels commented on src/uu/seq/src/seq.rs:
```diff
@@ -268,38 +200,64 @@ fn done_printing<T: Num + PartialOrd>(next: &T, increment: &T, last: &T) -> bool
-        i += 1;
-        value = first + i as f64 * increment;
+        write!(stdout, "{}", istr)?;
+        value += increment.clone();
```
I'll change that.

Does using a reference here avoid a copy somehow? I'm interested to learn more.

blyxxyz replied:
Probably. `BigDecimal` [implements](https://docs.rs/bigdecimal/0.3.0/bigdecimal/struct.BigDecimal.html#trait-implementations) both adding owned `BigDecimal`s and adding references, but I don't see a reason why it'd want to consume the other number, so I think adding owned numbers is just for convenience.

jfinkels commented:
I created separate pull requests for two of the suggested changes and made all of the suggested changes except:

> I think changing `RangeBigDecimal` into `(BigDecimal, BigDecimal, Option<BigDecimal>)` could work. With something like this logic: [...]

I'll try that next. Thanks.

jfinkels commented:
I'm going to approach this from a different angle. I'll open a new pull request when I have something. Thanks for the feedback on this.
